src/MongoDB.AspNetCore.OData/MongoEnableQueryAttribute.cs

```diff
@@ -0,0 +1,141 @@
+            }
+        }
+    }
+
+    public override IQueryable ApplyQuery(IQueryable queryable, ODataQueryOptions queryOptions)
+    {
+        if (queryable is IMongoQueryable)
+        {
+            var applyImplMethodInfo = s_applyQueryInternalMethodInfo.MakeGenericMethod(queryable.ElementType);
+            return (IQueryable)applyImplMethodInfo.Invoke(null, new object[] { queryable, queryOptions });
+        }
+
+        return base.ApplyQuery(queryable, queryOptions);
+    }
+
+    private static IQueryable ApplyQueryInternal<T>(IQueryable<T> queryable, ODataQueryOptions queryOptions)
+    {
+        var ignoreQueryOptions = AllowedQueryOptions.Select | AllowedQueryOptions.Expand;
+        queryable = (IQueryable<T>)queryOptions.ApplyTo(queryable, ignoreQueryOptions);
+
+        if (queryOptions.Count?.Value == true)
+        {
+            return queryable;
+        }
+
+        queryable = ApplyProjection(queryable, queryOptions);
+        // SelectExpand projection generated by oData library is not compatible with Mongo LINQ provider,
+        // so we have to execute the query before applying it. And then reapply Select and Expand to in-memory objects.
+        // It should be done here to turn IMongoQueryable into IEnumerable, so the detection in OnActionExecuted works as expected.
+        queryable = ExecuteQuery(queryable);
+
+        if (queryOptions.SelectExpand != null)
+        {
+            var allowedOptions = ignoreQueryOptions | AllowedQueryOptions.Compute;
+            var ignoreOption = AllowedQueryOptions.All & ~allowedOptions;
+            return queryOptions.ApplyTo(queryable, ignoreOption);
+        }
+
+        return queryable;
+    }
+
+    private static IQueryable<T> ApplyProjection<T>(IQueryable<T> queryable, ODataQueryOptions queryOptions)
+    {
+        var fieldProjects = new List<ProjectionDefinition<T>>();
+        var entityType = queryOptions.Context.NavigationSource.EntityType();
+
+        if (queryOptions.Compute != null || queryOptions.SelectExpand == null || queryOptions.SelectExpand.SelectExpandClause.AllSelected)
+        {
+            fieldProjects.Add(entityType.StructuralProperties().Project<T>());
+        }
+
+        if (queryOptions.SelectExpand != null)
+        {
+            foreach (var selectedItem in queryOptions.SelectExpand.SelectExpandClause.SelectedItems)
+            {
+                var translator = new MongoProjectionSelectItemTranslator<T>(queryOptions.Context);
+                fieldProjects.Add(selectedItem.TranslateWith(translator));
+            }
+        }
+
+        var projection = fieldProjects.Combine();
+        var projectionStage = PipelineStageDefinitionBuilder.Project<T, T>(projection);
+        return ((IMongoQueryable<T>)queryable).AppendStage(projectionStage);
+    }
+
+    private static IQueryable<T> ExecuteQuery<T>(IQueryable<T> queryable)
```

Comment:
I don't have good suggestions, maybe `ChangeQueryProvider`?

Reply:
As it was discussed the method was removed at all =)